File: knowledge-base-frontend-master/src/services/statistics.service.ts
import dayjs from 'dayjs';
import { http } from './request';
import { SystemStatistics, DashboardStats } from '@/types';

// Compute the date range based on the period
const getDateRange = (period: 'week' | 'month' | 'year'): { startDate: string; endDate: string } => {
  const endDate = dayjs().format('YYYY-MM-DD');
  const dayMap = { week: 6, month: 29, year: 364 };
  const startDate = dayjs().subtract(dayMap[period], 'days').format('YYYY-MM-DD');
  return { startDate, endDate };
};

export const statisticsService = {
  // Get system statistics overview
  getSystemStatistics: (params?: { startDate?: string; endDate?: string }) => {
    return http.get<SystemStatistics>('/statistics/overview', { params });
  },

  // Get dashboard statistics
  getDashboardStats: () => {
    return http.get<DashboardStats>('/statistics/dashboard');
  },

  // Get admin dashboard overview
  getAdminOverview: () => {
    return http.get('/statistics/admin-overview');
  },

  // Get document trend
  getDocumentTrend: (params: { period: 'week' | 'month' | 'year' }) => {
    const { startDate, endDate } = getDateRange(params.period);
    return http.get<Array<{ date: string; count: number }>>('/statistics/trend/document', {
      params: { startDate, endDate, type: 'create' },
    });
  },

  // Get user activity
  getUserActivity: (params: { period: 'week' | 'month' | 'year' }) => {
    const { startDate, endDate } = getDateRange(params.period);
    return http.get<Array<{ date: string; count: number }>>('/statistics/activity/user', {
      params: { startDate, endDate },
    });
  },

  // Get category distribution
  getCategoryDistribution: () => {
    return http.get<Array<{ name: string; value: number }>>('/statistics/distribution/category');
  },

  // Get popular documents (composite popularity ranking)
  getPopularDocuments: (params?: { limit?: number; period?: 'week' | 'month' | 'all' }) => {
    return http.get<Array<{
      documentId: string;
      title: string;
      authorName?: string;
      categoryName?: string;
      viewCount: number;
      likeCount: number;
      favoriteCount: number;
      summary?: string;
      createdAt?: string;
    }>>(
      '/statistics/hot/document',
      { params: { type: 'composite', size: params?.limit || 6 } }
    );
  },

  // Get latest documents
  getLatestDocuments: (params?: { limit?: number }) => {
    return http.get<Array<{
      documentId: string;
      title: string;
      authorName?: string;
      categoryName?: string;
      viewCount: number;
      likeCount: number;
      favoriteCount: number;
      summary?: string;
      createdAt?: string;
    }>>(
      '/statistics/latest/documents',
      { params: { size: params?.limit || 6 } }
    );
  },

  // Get active users
  getActiveUsers: (params?: { limit?: number }) => {
    return http.get<Array<{ userId: string; username: string; contribution: number }>>(
      '/statistics/active/user',
      { params: { type: 'create', size: params?.limit || 10 } }
    );
  },
};

export default statisticsService;

File: knowledge-base-backend-master/kb-statistics/src/main/java/com/knowledge/base/statistics/service/StatisticsService.java
package com.knowledge.base.statistics.service;

import com.knowledge.base.statistics.vo.*;

import java.time.LocalDate;
import java.util.List;

/**
 * Statistics analysis service interface
 *
 * <p>Provides business logic related to data statistics</p>
 *
 * @author airwzz999
 * @since 1.0.0
 */
public interface StatisticsService {

    /**
     * Gets the data overview
     *
     * @return the data overview information
     */
    OverviewVO getOverview();

    /**
     * Gets the document trend
     *
     * @param startDate the start date
     * @param endDate   the end date
     * @param type      the trend type (create, view, like, favorite)
     * @return the list of trend data
     */
    List<TrendVO> getDocumentTrend(LocalDate startDate, LocalDate endDate, String type);

    /**
     * Gets user activity
     *
     * @param startDate the start date
     * @param endDate   the end date
     * @return the list of activity data
     */
    List<UserActivityVO> getUserActivity(LocalDate startDate, LocalDate endDate);

    /**
     * Gets the category distribution
     *
     * @return the list of category distribution data
     */
    List<CategoryDistributionVO> getCategoryDistribution();

    /**
     * Gets popular documents
     *
     * @param type the statistics type (view, like, favorite)
     * @param size the count
     * @return the list of popular documents
     */
    List<HotDocumentVO> getHotDocuments(String type, Integer size);

    /**
     * Gets active users
     *
     * @param type the statistics type (create, comment, view)
     * @param size the count
     * @return the list of active users
     */
    List<ActiveUserVO> getActiveUsers(String type, Integer size);

    /**
     * Gets the dashboard data
     *
     * @return the comprehensive dashboard data
     */
    DashboardVO getDashboardData();

    /**
     * Gets the latest documents
     *
     * @param size the count
     * @return the latest document list
     */
    List<HotDocumentVO> getLatestDocuments(Integer size);

    /**
     * Gets the admin overview
     *
     * @return the admin overview data
     */
    AdminOverviewVO getAdminOverview();
}
